### PR TITLE
Remove code duplication in reviews data source

### PR DESCRIPTION
**Pull request by toupper**
Part of: #6523

### Description

The logic for showing reviews in the Products and Menu tabs is almost entirely the same, but still, we have duplicated classes for it. With this Pull Request, we start refactoring it by having only one `ReviewsDataSource` instance, injecting the different behavior for each case via `ReviewsDataSourceCustomizing`. With that, we will have 167 lines less of code.

### Notes
After these changes, we will have a protocol with just one implementation. To avoid blowing up the size of this PR I didn't add it here, but In the next PR I will refactor to use the naming convention for this case: `ReviewsDataSourceProtocol` and `ReviewsDataSource`.

### Changes
- Extract different logic into a protocol, `ReviewsDataSourceCustomizing`
- Implement this protocol for the Products and Menu tabs (`ProductReviewsDataSourceCustomizer` and `GlobalReviewsDataSourceCustomizer`)
- Use this injected logic in `DefaultReviewsDataSource`
- Use the `DefaultReviewsDataSource` for products as well
- Remove `ProductsReviewsDataSource`
- Unit Tests

#### Products

1. Open App
2. Go to Products tab
3. Go to a product with reviews
4. Open reviews 
5. Open a review

The flow and design should be the same as before

#### Menu

1. Open App
2. Go to Menu tab
3. Open reviews
4. Open a review

The flow and design should be the same as before

#### Products
reviews-refactor-products
#### Menu

review-refactor-menu

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by pmusolino — COMMENTED** [review]
While I was testing this with the reviews of one of my products (`Italian Pizza`), the list of reviews was empty (in reality it's not), and when I tried to do a pull to refresh action, I encountered this crash:

```
Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'attempt to delete row 2 from section 0, but there are only 0 sections before the update'
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001803e1188 __exceptionPreprocess + 236
	1   libobjc.A.dylib                     0x0000000180193384 objc_exception_throw + 56
	2   Foundation                          0x0000000180746648 _userInfoForFileAndLine + 0
	3   UIKitCore                           0x0000000184fa09a8 -[UITableView _endCellAnimationsWithContext:] + 5324
	4   UIKitCore                           0x0000000184fba5bc -[UITableView endUpdatesWithContext:] + 140
	5   WooCommerce                         0x0000000104a682a0 $s8Yosemite17ResultsControllerC11WooCommerceE28startForwardingContentEvents33_B67F1DDE16B1959152AC062605FC5A2DLL2to4withySo11UITableViewC_AD0B15TableAnimationsVtFyycfU0_ + 140
	6   Yosemite                            0x000000010c65be34 $s8Yosemite17ResultsControllerC21setupEventsForwarding33_859B47074559A83DD5AD581E243E51FELLyyFyycfU0_ + 240
	7   Yosemite                            0x000000010c66bd8c $s8Yosemite39FetchedResultsControllerDelegateWrapperC26controllerDidChangeContentyySo09NSFetchedcD0CySo20NSFetchRequestResult_pGF + 124
	8   Yosemite                            0x000000010c66bde4 $s8Yosemite39FetchedResultsControllerDelegateWrapperC26controllerDidChangeContentyySo09NSFetchedcD0CySo20NSFetchRequestResult_pGFTo + 52
	9   CoreData                            0x00000001858f1ea4 __82-[NSFetchedResultsController(PrivateMethods) _core_managedObjectContextDidChange:]_block_invoke + 6576
	10  CoreData                            0x0000000185885308 developerSubmittedBlockToNSManagedObjectContextPerform + 156
	11  CoreData                            0x00000001858851dc -[NSManagedObjectContext performBlockAndWait:] + 216
	12  CoreData                            0x00000001858f04d8 -[NSFetchedResultsController _core_managedObjectContextDidChange:] + 112
	13  CoreFoundation                      0x0000000180321404 __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 20
	14  CoreFoundation                      0x00000001803213bc ___CFXRegistrationPost_block_invoke + 48
	15  CoreFoundation                      0x00000001803207e0 _CFXRegistrationPost + 416
	16  CoreFoundation                      0x00000001803201c8 _CFXNotificationPost + 680
	17  Foundation                          0x00000001807c442c -[NSNotificationCenter postNotificationName:object:userInfo:] + 92
	18  CoreData                            0x00000001858761c0 -[NSManagedObjectContext _postObjectsDidChangeNotificationWithUserInfo:] + 436
	19  CoreData                            0x000000018588b838 -[NSManagedObjectContext _createAndPostChangeNotification:deletions:updates:refreshes:deferrals:wasMerge:] + 1572
	20  CoreData                            0x0000000185878148 -[NSManagedObjectContext _processRecentChanges:] + 3172
	21  CoreData                            0x0000000185890e68 -[NSManagedObjectContext _parentProcessSaveRequest:inContext:error:] + 1268
	22  CoreData                            0x0000000185891440 __82-[NSManagedObjectContext(_NestedContextSupport) executeRequest:withContext:error:]_block_invoke + 564
	23  CoreData                            0x0000000185893b3c internalBlockToNSManagedObjectContextPerform + 104
	24  libdispatch.dylib                   0x0000000107f51694 _dispatch_client_callout + 16
	25  libdispatch.dylib                   0x0000000107f61334 _dispatch_async_and_wait_invoke + 192
	26  libdispatch.dylib                   0x0000000107f51694 _dispatch_client_callout + 16
	27  libdispatch.dylib                   0x0000000107f601fc _dispatch_main_queue_callback_4CF + 1084
	28  CoreFoundation                      0x000000018034fcd4 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 12
	29  CoreFoundation                      0x000000018034a244 __CFRunLoopRun + 2448
	30  CoreFoundation                      0x00000001803493a8 CFRunLoopRunSpecific + 572
	31  GraphicsServices                    0x000000018c03c5ec GSEventRunModal + 160
	32  UIKitCore                           0x0000000184d937ac -[UIApplication _run] + 992
	33  UIKitCore                           0x0000000184d982e8 UIApplicationMain + 112
	34  WooCommerce                         0x0000000104b18940 main + 452
	35  dyld                                0x00000001075a5ca0 start_sim + 20
	36  ???                                 0x00000001076650f4 0x0 + 4419113204
	37  ???                                 0xf926800000000000 0x0 + 17953177702047416320
)
libc++abi: terminating with uncaught exception of type NSException
dyld4 config: DYLD_ROOT_PATH=/Applications/Xcode-13.2.1.app/Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot DYLD_LIBRARY_PATH=/Users/pmusolino/Library/Developer/Xcode/DerivedData/WooCommerce-fnqvgxzzgmalhtcworiuipfeecdt/Build/Products/Debug-iphonesimulator:/Applications/Xcode-13.2.1.app/Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot/usr/lib/system/introspection DYLD_INSERT_LIBRARIES=/Applications/Xcode-13.2.1.app/Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot/usr/lib/libBacktraceRecording.dylib:/Applications/Xcode-13.2.1.app/Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot/usr/lib/libMainThreadChecker.dylib:/Applications/Xcode-13.2.1.app/Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot/usr/lib/libMTLCapture.dylib:/Applications/Xcode-13.2.1.app/Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot/Developer/Library/PrivateFrameworks/DTDDISupport.framework/libViewDebuggerSupport.dylib DYLD_FRAMEWORK_PATH=/Users/pmusolino/Library/Developer/Xcode/DerivedData/WooCommerce-fnqvgxzzgmalhtcworiuipfeecdt/Build/Products/Debug-iphonesimulator:/Users/pmusolino/Library/Developer/Xcode/DerivedData/WooCommerce-fnqvgxzzgmalhtcworiuipfeecdt/Build/Products/Debug-iphonesimulator/PackageFrameworks
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'attempt to delete row 2 from section 0, but there are only 0 sections before the update'
terminating with uncaught exception of type NSException
CoreSimulator 802.6 - Device: iPhone 13 Pro Max (3DC9257C-1822-48EB-9D34-57D16E649E08) - Runtime: iOS 15.2 (19C51) - DeviceType: iPhone 13 Pro Max
(lldb) 
```

I tried to replicate this after the crash. During the first loading the view was still empty. I tried a pull to refresh and the data now loaded correctly, and I wasn't able to experience anymore the crash (I suspect that we need a clean installation).

**Comment by toupper** [comment]
@pmusolino Thanks for your review! Actually, that problem and crash also happen on trunk, so I created [this issue ](https://github.com/woocommerce/woocommerce-ios/issues/6568) and will take care of it now. Afterward, I will merge the solution to this PR as well 👍

**Comment by pmusolino** [comment]
> @pmusolino Thanks for your review! Actually, that problem and crash also happen on trunk, so I created https://github.com/woocommerce/woocommerce-ios/issues/6568and will take care of it now. Afterward, I will merge the solution to this PR as well

I confirm that it happens also on `trunk` 👍

File: RELEASE-NOTES.txt [hunk 1]
```diff
@@ -1,5 +1,9 @@
 *** PLEASE FOLLOW THIS FORMAT: [<priority indicator, more stars = higher priority>] <description> [<PR URL>]
 
+9.0
+-----
+- [internal] Reviews lists on Products and Menu tabs are refactored to avoid duplicated code. Please quickly smoke test them to make sure that everything still works as before. [https://github.com/woocommerce/woocommerce-ios/pull/6553]
```
Comment (pmusolino) [inline RELEASE-NOTES.txt:5]: 💯

**Review by pmusolino — APPROVED** [review]
Looks good to me, great job @toupper 🥇 
It's always good to see some code removed 😄

**Pull request merged**